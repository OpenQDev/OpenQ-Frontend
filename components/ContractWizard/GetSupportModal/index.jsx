import Image from 'next/image';
import React, { useEffect, useRef } from 'react';
import ModalDefault from '../../Utils/ModalDefault';

const GetSupportModal = ({ wizardVisibility, modalVisibility }) => {
  // Refs
  const modal = useRef();

  useEffect(() => {
    // Courtesy of https://stackoverflow.com/questions/32553158/detect-click-outside-react-component
    function handleClickOutside(event) {
      if (modal.current && !modal.current.contains(event.target)) {
        modalVisibility(false);
        wizardVisibility(false);
      }
    }
    // Bind the event listener
    {
      document.addEventListener('mousedown', handleClickOutside);
    }
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [modal]);

  const resetState = () => {
    modalVisibility(false);
  };

  const btn = (
    <a
      href={'https://discord.gg/puQVqEvVXn'}
      className='flex items-center gap-2 btn-primary'
      target={'_blank'}
      rel='noopener noreferrer'
    >
      <div data-testid='link' className='flex items-center gap-2'>
        <Image src={'/social-icons/discord.svg'} width={24} height={24} alt='discord-logo' />
        Get Help!
      </div>
    </a>
  );

  return (
    <>
      <ModalDefault
        title={"We didn't find a suitable contract"}
        footerRight={btn}
        setShowModal={wizardVisibility}
        resetState={resetState}
      >
        <h3 className='text-lg pt-4'>
          Don't worry, you can help us and describe more precisely what kind of contract you want. If we don't have this
          one yet, we'll be happy to get to it next.
        </h3>
        <p className='pt-4 pr-4'>
          Join our Discord and write us in the support channel or send us an email at riccardo@openq.dev
        </p>
      </ModalDefault>
    </>
  );
};

export default GetSupportModal;
