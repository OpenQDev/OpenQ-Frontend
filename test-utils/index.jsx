// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
