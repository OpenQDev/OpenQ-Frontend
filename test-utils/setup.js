import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
