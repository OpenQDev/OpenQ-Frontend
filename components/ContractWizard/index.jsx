import React, { useEffect, useRef, useState } from 'react';
import MintBountyModal from '../MintBounty/MintBountyModal/MintBountyModal';
import MintProvider from '../MintBounty/MintProvider';
import ModalDefault from '../Utils/ModalDefault';
import ToolTipNew from '../Utils/ToolTipNew';
import GetSupportModal from './GetSupportModal';

const ContractWizard = ({ wizardVisibility }) => {
  const [type, setType] = useState(0);
  const [mintModal, setMintModal] = useState(false);
  const [supportModal, setSupportModal] = useState(false);
  const [mintActivated, setMintActivated] = useState(false);
  // Refs
  const modal = useRef();

  // Methods
  function handleYes() {
    setMintModal(true);
    setMintActivated(true);
  }

  useEffect(() => {
    // Courtesy of https://stackoverflow.com/questions/32553158/detect-click-outside-react-component
    function handleClickOutside(event) {
      if (modal.current && !modal.current.contains(event.target)) {
        wizardVisibility(false);
      }
    }
    // Bind the event listener
    {
      document.addEventListener('mousedown', handleClickOutside);
    }
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [modal, mintModal, supportModal]);

  useEffect(() => {
    if (type > 3) {
      setSupportModal(true);
    }
  }, [type]);

  useEffect(() => {
    if (mintActivated && !mintModal) {
      wizardVisibility(false);
    }
  }, [mintModal]);

  const handleTypeChange = () => {
    if (type == 0) {
      setType(type + 3);
    } else setType(type + 1);
  };
  const resetState = () => {
    wizardVisibility(false);
  };

  const texts = new Map([
    [
      0,
      [
        'Should only one person complete this task?',
        'Deploy a Fixed Price Contract to fund a single issue to be payed out to one submitter.',
      ],
    ],
    [
      1,
      [
        'Should several people be able to earn the same amount of money for this task? This contract is especially suitable for Learn 2 Earn.',
        'Deploy a Split Price Contract, where several people can submit and claim the funds for the same issue.',
      ],
    ],
    [
      2,
      [
        'Do you want to create a contest and allow several people to earn prizes? In this case you can set different weights and choose multiple winners.',
        'Particulary suitable for Hackathons or other contests.',
      ],
    ],
    [
      3,
      [
        'Do you want to create a contest and allow several people to earn fixed prizes? In this case you can set fixed prizes and choose multiple winners.',
        'Particulary suitable for Hackathons or other contests.',
      ],
    ],
  ]);
  const text = (type) => {
    return texts.get(type) || ['Type of contract unkown.', '?'];
  };

  const btn = (
    <div className='flex text-sm rounded-sm overflow-hidden w-fit text-primary '>
      <button
        onClick={handleYes}
        className='w-fit min-w-[80px] py-[5px] px-4 rounded-l-sm border whitespace-nowrap hover:bg-secondary-button hover:border-secondary-button border-web-gray'
      >
        Yes
      </button>
      <button
        onClick={handleTypeChange}
        className='w-fit min-w-[80px] py-[5px] px-4 border-l-0 rounded-r-sm border whitespace-nowrap hover:bg-secondary-button hover:border-secondary-button border-web-gray'
      >
        No
      </button>
    </div>
  );
  return (
    <>
      {!mintModal ? (
        !supportModal ? (
          <ModalDefault
            title={'Contract Wizard'}
            footerRight={btn}
            setShowModal={wizardVisibility}
            resetState={resetState}
          >
            <h3 className='text-2xl pt-2'>
              In this section we will help you find the right type of contract for your job.
            </h3>
            <div className='flex items-center gap-2 pt-4'>
              {text(type)[0]}
              <ToolTipNew
                relativePosition={'-right-5'}
                outerStyles={'relative bottom-1'}
                innerStyles={'whitespace-normal w-60'}
                toolTipText={text(type)[1]}
              >
                <div className='cursor-help rounded-full border border-[#c9d1d9] aspect-square leading-4 h-4 box-content text-center font-bold text-primary'>
                  ?
                </div>
              </ToolTipNew>
            </div>
          </ModalDefault>
        ) : (
          <GetSupportModal wizardVisibility={wizardVisibility} modalVisibility={setSupportModal} />
        )
      ) : (
        <>
          {mintModal && (
            <MintProvider type={type}>
              <MintBountyModal modalVisibility={setMintModal} />
            </MintProvider>
          )}
        </>
      )}
    </>
  );
};

export default ContractWizard;
