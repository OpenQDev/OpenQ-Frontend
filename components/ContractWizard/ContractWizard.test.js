/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '../../test-utils';
import ContractWizard from '.';
import userEvent from '@testing-library/user-event';
import ShallowRenderer from 'react-test-renderer/shallow';
import nextRouter from 'next/router';

nextRouter.useRouter = jest.fn();
nextRouter.useRouter.mockImplementation(() => ({
  query: { type: null },

  prefetch: jest.fn(() => {
    return { catch: jest.fn };
  }),
}));

describe('ContractWizard', () => {
  it('should render match DOM Snapshot', () => {
    const shallow = new ShallowRenderer();
    shallow.render(<ContractWizard wizardVisibility={true} />);
    const tree = shallow.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should open wizard and direct to discord server', async () => {
    // ARRANGE
    const user = userEvent.setup();
    render(<ContractWizard wizardVisibility={true} />);

    // ACT
    await user.click(screen.getByText('No'));
    await user.click(screen.getByText('No'));
    expect(await screen.findByText(/we didn't find a suitable contract/i)).toBeInTheDocument();
  });

  it('should open wizard and direct to fixed contest contract', async () => {
    // ARRANGE
    const user = userEvent.setup();
    render(<ContractWizard wizardVisibility={true} />);

    // ACT
    await user.click(screen.getByText('No'));
    await user.click(screen.getByText('Yes'));
    expect(await screen.findByText(/Create a Hackathon Contract to send funds to any GitHub issue/i));
  });
});
