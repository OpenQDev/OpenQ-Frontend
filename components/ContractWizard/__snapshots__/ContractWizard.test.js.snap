// vi Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractWizard should render match DOM Snapshot 1`] = `
<React.Fragment>
  <ModalDefault
    footerRight={
      <div
        className="flex text-sm rounded-sm overflow-hidden w-fit text-primary "
      >
        <button
          className="w-fit min-w-[80px] py-[5px] px-4 rounded-l-sm border whitespace-nowrap hover:bg-secondary-button hover:border-secondary-button border-web-gray"
          onClick={[Function]}
        >
          Yes
        </button>
        <button
          className="w-fit min-w-[80px] py-[5px] px-4 border-l-0 rounded-r-sm border whitespace-nowrap hover:bg-secondary-button hover:border-secondary-button border-web-gray"
          onClick={[Function]}
        >
          No
        </button>
      </div>
    }
    resetState={[Function]}
    setShowModal={true}
    title="Contract Wizard"
  >
    <h3
      className="text-2xl pt-2"
    >
      In this section we will help you find the right type of contract for your job.
    </h3>
    <div
      className="flex items-center gap-2 pt-4"
    >
      Should only one person complete this task?
      <ToolTipNew
        innerStyles="whitespace-normal w-60"
        outerStyles="relative bottom-1"
        relativePosition="-right-5"
        toolTipText="Deploy a Fixed Price Contract to fund a single issue to be payed out to one submitter."
      >
        <div
          className="cursor-help rounded-full border border-[#c9d1d9] aspect-square leading-4 h-4 box-content text-center font-bold text-primary"
        >
          ?
        </div>
      </ToolTipNew>
    </div>
  </ModalDefault>
</React.Fragment>
`;
